// a11y-sitechecker/lib/utils/setup-config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupAxe = exports.setupAxeConfig = exports.prepareWorkspace = exports.setupConfig = void 0;
const fs_1 = __importDefault(require("fs"));
const helper_functions_1 = require("./helper-functions");
const puppeteer_1 = require("@axe-core/puppeteer");
function setupConfig(options) {
    const config = {
        json: true,
        resultsPath: 'results/',
        resultsPathPerUrl: '',
        axeConfig: {},
        threshold: 0,
        imagesPath: 'images',
        timeout: 30000,
        debugMode: false,
        viewports: [
            {
                width: 1920,
                height: 1080,
            },
        ],
        resultTypes: ['violations', 'incomplete'],
        runOnly: ['wcag2aa', 'wcag2a', 'wcag21a', 'wcag21aa', 'best-practice', 'ACT', 'experimental'],
        crawl: false,
        name: '',
        urlsToAnalyze: [],
    };
    config.json = options.json;
    if (options.threshold) {
        config.threshold = parseInt(options.threshold);
    }
    if (options.config || options.providedConfig) {
        try {
            let configFile;
            if (options.config) {
                configFile = JSON.parse(fs_1.default.readFileSync(options.config).toString('utf-8'));
            }
            else {
                configFile = options.providedConfig;
            }
            if (typeof configFile.json === 'boolean') {
                config.json = configFile.json;
            }
            if (configFile.resultsPath && typeof configFile.resultsPath === 'string') {
                config.resultsPath = configFile.resultsPath + (configFile.resultsPath.endsWith('/') ? '' : '/');
            }
            if (configFile.axeConfig?.locale && typeof configFile.axeConfig.locale === 'string' && config.axeConfig) {
                config.axeConfig.locale = configFile.axeConfig.locale;
            }
            if (configFile.axeConfig?.localePath &&
                typeof configFile.axeConfig.localePath === 'string' &&
                config.axeConfig) {
                config.axeConfig.localePath = configFile.axeConfig.localePath;
            }
            if (configFile.login) {
                config.login = configFile.login;
            }
            if (configFile.launchOptions) {
                config.launchOptions = configFile.launchOptions;
            }
            if (configFile.saveImages) {
                config.saveImages = configFile.saveImages;
            }
            if (configFile.ignoreElementAttributeValues) {
                config.ignoreElementAttributeValues = configFile.ignoreElementAttributeValues;
            }
            if (configFile.urlsToAnalyze) {
                config.urlsToAnalyze = configFile.urlsToAnalyze;
            }
            else {
                throw new Error('It is absolutly necessary to provide 1 or more urls!');
            }
            if (configFile.analyzeClicksWithoutNavigation) {
                config.analyzeClicksWithoutNavigation = configFile.analyzeClicksWithoutNavigation;
            }
            if (configFile.debugMode && typeof configFile.debugMode === 'boolean') {
                config.debugMode = configFile.debugMode;
            }
            if (configFile.analyzeClicks && typeof configFile.analyzeClicks === 'boolean') {
                config.analyzeClicks = configFile.analyzeClicks;
            }
            if (configFile.timeout && typeof configFile.timeout === 'number') {
                config.timeout = configFile.timeout;
            }
            if (configFile.viewports) {
                config.viewports = configFile.viewports;
            }
            if (configFile.resultTypes) {
                config.resultTypes = configFile.resultTypes;
            }
            if (configFile.runOnly) {
                config.runOnly = configFile.runOnly;
            }
            if (configFile.idTags) {
                config.idTags = configFile.idTags;
            }
            if (configFile.clickableItemSelector) {
                config.clickableItemSelector = configFile.clickableItemSelector;
            }
            if (configFile.cookieSelector) {
                config.cookieSelector = configFile.cookieSelector;
            }
            if (configFile.cookieText) {
                config.cookieText = configFile.cookieText;
            }
            if (configFile.crawl) {
                if (config.urlsToAnalyze?.length === 1 && configFile.crawl) {
                    config.crawl = true;
                }
                else {
                    throw new Error('Crawling is only possible with one Url supplied!');
                }
            }
            if (configFile.name) {
                config.name = configFile.name;
            }
            else {
                throw new Error('It is absolutly necessary to provide a name!');
            }
            if (configFile.threshold) {
                config.threshold = configFile.threshold;
            }
            if (configFile.screenshotPadding) {
                config.screenshotPadding = configFile.screenshotPadding;
            }
        }
        catch (e) {
            (0, helper_functions_1.error)(e);
            throw e;
        }
    }
    else {
        throw new Error('It is absolutly necessary to provide a config!');
    }
    return config;
}
exports.setupConfig = setupConfig;
function prepareWorkspace(config) {
    const urlEscaped = (0, helper_functions_1.getEscaped)(config.name);
    config.resultsPathPerUrl = config.resultsPath + urlEscaped + '/';
    config.imagesPath = config.resultsPathPerUrl + 'images/';
    if (config.imagesPath && !fs_1.default.existsSync(config.imagesPath) && config.saveImages) {
        fs_1.default.mkdirSync(config.imagesPath, { recursive: true });
    }
    else if (config.imagesPath && config.saveImages) {
        fs_1.default.rmSync(config.imagesPath, { recursive: true });
        fs_1.default.mkdirSync(config.imagesPath, { recursive: true });
    }
    if (config.resultsPathPerUrl && !fs_1.default.existsSync(config.resultsPathPerUrl)) {
        fs_1.default.mkdirSync(config.resultsPathPerUrl, { recursive: true });
    }
}
exports.prepareWorkspace = prepareWorkspace;
function setupAxeConfig(config) {
    const axeConfig = {};
    try {
        if (config.axeConfig?.locale) {
            axeConfig.locale = JSON.parse(fs_1.default
                .readFileSync('./node_modules/axe-core/locales/' + config.axeConfig.locale + '.json')
                .toString('utf-8'));
        }
        else if (config.axeConfig?.localePath) {
            axeConfig.locale = JSON.parse(fs_1.default.readFileSync(config.axeConfig.localePath).toString('utf-8'));
        }
    }
    catch (e) {
        (0, helper_functions_1.error)('Locale not found. Using Standard locale "en"');
    }
    return axeConfig;
}
exports.setupAxeConfig = setupAxeConfig;
async function setupAxe(page, axeSpecs, config) {
    const axe = new puppeteer_1.AxePuppeteer(page);
    axe.configure(axeSpecs);
    axe.options({
        runOnly: config.runOnly,
        resultTypes: config.resultTypes,
    });
    return axe;
}
exports.setupAxe = setupAxe;
