// a11y-sitechecker/lib/utils/setup-siteresult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupSiteresult = void 0;
function setupSiteresult(id, sitecheckerResult) {
    return {
        id: id,
        name: sitecheckerResult.name,
        toolOptions: sitecheckerResult.toolOptions,
        testEngine: sitecheckerResult.testEngine,
        testRunner: sitecheckerResult.testRunner,
        testEnvironment: sitecheckerResult.testEnvironment,
        timestamp: sitecheckerResult.timestamp,
        analyzedUrls: sitecheckerResult.analyzedUrls,
        countViolations: sitecheckerResult.violations
            .map((v) => v.nodes.length)
            .reduce((count, value) => count + value, 0),
        countPasses: sitecheckerResult.passes.map((v) => v.nodes.length).reduce((count, value) => count + value, 0),
        countIncomplete: sitecheckerResult.incomplete
            .map((v) => v.nodes.length)
            .reduce((count, value) => count + value, 0),
        countInapplicable: sitecheckerResult.inapplicable.length,
        tabables: sitecheckerResult.tabables,
    };
}
exports.setupSiteresult = setupSiteresult;
